gxa/src/main/java/uk/ac/ebi/atlas/experimentpage/differential/DifferentialPathwaysComparisonService.java

```diff
@@ -0,0 +1,133 @@
+import uk.ac.ebi.atlas.model.experiment.differential.*;
+import uk.ac.ebi.atlas.resource.DataFileHub;
+import uk.ac.ebi.atlas.search.SemanticQuery;
+import uk.ac.ebi.atlas.trader.ConfigurationTrader;
+import uk.ac.ebi.atlas.web.DifferentialRequestPreferences;
+
+import java.util.*;
+
+import static java.util.stream.Collectors.toMap;
+
+public class DifferentialPathwaysComparisonService <Expr extends DifferentialExpression, E extends DifferentialExperiment,
+        K extends DifferentialRequestPreferences,
+        P extends DifferentialProfile<Expr, P>,  R extends DifferentialRequestContext<E, K> >
+        extends ExperimentPageService {
+
+    private final DataFileHub dataFileHub;
+    private final ConfigurationTrader configurationTrader;
+    private final DifferentialRequestContextFactory<E, K, R> differentialRequestContextFactory;
+    private final DifferentialProfilesHeatMap<Expr, E, P, R> profilesHeatMap;
+
+
+    public DifferentialPathwaysComparisonService(DataFileHub dataFileHub,
+                                                 ConfigurationTrader configurationTrader,
+                                                 DifferentialRequestContextFactory<E, K, R> differentialRequestContextFactory,
+                                                 DifferentialProfilesHeatMap<Expr, E, P, R> profilesHeatMap) {
+
+        this.differentialRequestContextFactory = differentialRequestContextFactory;
+        this.profilesHeatMap = profilesHeatMap;
+        this.dataFileHub = dataFileHub;
+        this.configurationTrader = configurationTrader;
+    }
+
+    public  Map<Contrast, List<String>> getPathwaysListPerComparison(DifferentialExperiment experiment) {
+        Map<Contrast, List<String>> comparisonsPathwaysMap = new HashMap<>();
+
+        String experimentAccession = experiment.getAccession();
+        List<Contrast> comparisons = configurationTrader.getMicroarrayExperimentConfiguration(experimentAccession).getContrasts();
+
+        //fetch pathwaysId from file for each comparison
+        for (Contrast comparison : comparisons) {
+            List<String> pathwaysValues = fetchPathwaysFromFile(experimentAccession, comparison);
+
+            comparisonsPathwaysMap.put(comparison, pathwaysValues);
+        }
+
+        return comparisonsPathwaysMap;
+    }
+
+    private List<String> fetchPathwaysFromFile (String experimentAccession, Contrast comparison) {
+        Map<String, Double> result = new HashMap<>();
+
+        List<String[]> lines = dataFileHub.getReactomePathwaysCFiles(experimentAccession, comparison.getId()).reactomePathways.get().readAll();
+
+        for(int i = 1; i < lines.size(); i++) {
+            String[] strings = lines.get(i);
+            String pathway = strings[0];
+            Double pvalue = Double.valueOf(strings[4]);
+
+            if (pvalue < 0.05) {
+                result.put(pathway, pvalue);
+            }
+        }
+
+        Map<String, Double> ordered = result.entrySet()
```

Comment:
You could do this here:
```
return result.entrySet().stream().sorted(Map.Entry.comparingByValue()).limit(10).map(Map.Entry::getKey).collect(Collectors.toList())
```

Reply:
I got it from the java 8 documentation for sorting the entrySet, and don't see anything wrong there

Reply:
You're making a new LinkedHashMap (with no reason that it should be precisely the LinkedHashMap), and then you're taking the keys of that hashmap, throwing away the values. The above is doing the same thing without the funky unnecessary part.

Reply:
Ok done! I made the last change you proposed